Add additional parameters into WC tracker for Pinterest

message-dimke opened the pull request:
### Changes proposed in this Pull Request:

- Adding additional WC tracker data: feed generation time and products count, store connection status and product sync status;
- Refactoring feed file operations our of feed generator to unit test feed generation failures and proper tracker snapshot properties update;
- Adding more unit tests on involved classes;

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

#### Basic test
1. Navigate to `WooCommerce > Status > Scheduled Actions > Pending`.
2. Run `pinterest-for-woocommerce-start-feed-generation` action.
3. Using WP CLI run a command `wp wc tracker snapshot`4. 
4. Check the comment output for 
```
... "store":{"connected":"yes","actively_syncing":"yes"},"feed":{"generation_time":111,"products_count":20} ...
```
5. `store` and `feed` objects must be present (they were not before, added within current PR) and have properties with values of the store connection status and sync status as well as recent feed generation details: time it took to generate the feed and the number of products went into the feed.
#### Test when previous results are displayed if feed generation is in progress
-1. Add new simple product into your catalog.
0. Put exit code above the line 173 to emulate feed generation job in progress (like it is having hard times generating 1100500 products)

https://github.com/woocommerce/pinterest-for-woocommerce/blob/0f0da243ff1ceec15b3f90fe029a0bedc9a9566b/src/FeedGenerator.php#L166-L173

**p.s.** _you can also add `sleep(3);` for example, because feed generation may run less than a second an wall time can have 0. value, by adding `sleep(3);` you know that wall time must be at least 3 and check it._

```
$this->feed_file_operations->rename_temporary_feed_files_to_final();
ProductFeedStatus::set(
	array(
		'status'                                                     => 'generated',
		ProductFeedStatus::PROP_FEED_GENERATION_RECENT_PRODUCT_COUNT => ProductFeedStatus::get()['product_count'],
	)
);
exit;
sleep(3);
ProductFeedStatus::set_feed_file_generation_time( time() );
```
1. Navigate to `WooCommerce > Status > Scheduled Actions > Pending`.
2. Run `pinterest-for-woocommerce-start-feed-generation` action.
3. Using WP CLI run a command `wp wc tracker snapshot`
4. Check the comment output for 
```
... "store":{"connected":"yes","actively_syncing":"yes"},"feed":{"generation_time":111,"products_count":20} ...
```
5. `store` and `feed` objects must be present and have the values from the previous **Basic test** case.
6. Go back to the code and remove `exit;`
7. Navigate to `WooCommerce > Status > Scheduled Actions > In progress` and remove in progress action on feed generation chain.
8. Navigate to `WooCommerce > Status > Scheduled Actions > Pending`.
9. Run `pinterest-for-woocommerce-start-feed-generation` action again.
10. Using WP CLI run a command `wp wc tracker snapshot`
11. Check the comment output for 
```
... "store":{"connected":"yes","actively_syncing":"yes"},"feed":{"generation_time":111,"products_count":20} ...
```
12. `feed` object must be present and contain the values from the new feed generation run: generation time at least 3, product count to be one product more that it was before.



### Changelog entry

> Add - Plugin connection and feed registration status tracking.
> Add - Recent feed generation time and feed product count tracking.

message-dimke committed:
Adding additional parameters into WC tracker for Pinterest: feed generation time, connection and sync statuses, feed products count.

message-dimke committed:
Fix phpcs short array syntax error.

message-dimke committed:
Refactor feed generator test.

message-dimke committed:
Fix php version syntax compatibility issues and unit test condition.

message-dimke committed:
Fix wrong test assertion parameters order.

message-dimke committed:
Fix add bool to string conversion on tracker parameter values; fix test wrong data key check.

message-dimke committed:
Refactor feed generation time calculation through product feed status instead of tracker snapshot object.

message-dimke committed:
Fix bugs with test suit, fix php code styling issues.

message-dimke committed:
Remove redundant constants.

budzanowski commented on src/TrackerSnapshot.php:
```diff
@@ -57,8 +56,18 @@ protected static function include_snapshot_data( array $data = array() ): array
+			),
+			'feed'     => array(
+				'generation_time' => $feed_generation_time,
+				'products_count'  => (int) ProductFeedStatus::get()['product_count'] ?? 0,
```
I think that the proper form is `product_count` and not `products...`.

budzanowski commented:
This looks good but I think that some modifications will be required.

The code will work correctly as long as the `WC_Tracker` snapshot will not happen during the feed generation process.
Since feed generation can take a lot of time, this can be a common occurrence.

This will lead to two things. The `PROP_FEED_GENERATION_WALL_TIME` will be equal to `0` and the `product_count` will be the number of products that have been processed since the start of the feed generation time.

budzanowski commented on src/FeedGenerator.php:
```diff
@@ -128,6 +128,8 @@ private function start_generation() {
 	 */
 	protected function handle_start() {
 		self::log( __( 'Feed generation start. Preparing temporary files.', 'pinterest-for-woocommerce' ) );
+		/* Remember the time a feed generation has started. */
```
Please use single line comments for singular sentences.

In general I would omit comments like this. Thats because you have used a very descriptive name.

budzanowski commented on src/ProductFeedStatus.php:
```diff
@@ -92,7 +98,47 @@ public static function set( $state ) {
 	}
+
+	/**
+	 * Resets a feed generation start time.
```
Please add `@since` to the methods docblock.
You can use `@since x.x.x` and the release process will automatically populate the version.

budzanowski commented:
Fantastic job with the Unit tests!

message-dimke committed:
Remove redundant comments.

message-dimke committed:
Add since and return phpDoc comments to the newly added methods.

message-dimke committed:
Rename tracker snapshot feed product count key.

message-dimke committed:
- Refactor feed file operations into separate class.
- Add unit tests on feed file operations class.
- Refactor feed generator class by removing file operation methods out of it.
- Refactor feed registration dependencies.
- Add recent product count property into product feed state storage to persist recent product count between feed generations and during the feed generation is in progress.
- Adjust tracker snapshot to get product count from the recent product count property.
- Tweak and extend unit test coverage.

message-dimke committed:
Adjust test method comment to explain what is the test about.

message-dimke commented:
Hey, @budzanowski! I did update PR description with two test cases and detailed instructions on how to run those tests. As well as edited more files and added more tests.

budzanowski commented on src/TrackerSnapshot.php:
```diff
@@ -57,8 +56,18 @@ protected static function include_snapshot_data( array $data = array() ): array
 			'settings' => self::parse_settings(),
+			'store'    => array(
+				'connected'        => wc_bool_to_string( Pinterest_For_Woocommerce::is_connected() ),
+				'actively_syncing' => wc_bool_to_string( ProductSync::is_product_sync_enabled() ),
```
This will be included in the `settings` section. It has `product_sync_enabled` prop tracked.
It would be better to track if the feed is registered. You can use `Pinterest_For_Woocommerce()::get_data( 'feed_registered' )` but don't store the ID just check if it is not null.

budzanowski commented on src/ProductFeedStatus.php:
```diff
@@ -90,9 +99,61 @@ public static function set( $state ) {
 	 */
 	public static function deregister() {
-
+		$properties_to_persist = array(
```
In case the feed is deregistered we should not send the count and time info - we should use the default values. This means that this logic can be eliminated.

budzanowski commented:
@message-dimke I have added some additional suggestions.

message-dimke committed:
Fix deregister must deregister all the product feed status properties. Fix tracker snapshot actively syncing value source.

message-dimke commented:
@budzanowski , if deregistered is it ok to show `time: 0` and` products: 0` and since `actively_syncing` will container value `'no'` we will know this is not a sync, but a default which means nothing?

If the above is good, then I did the changes and committed. Ready for another review round.

budzanowski commented:
I think that it is OK.
